### PR TITLE
[AOT] Add middleware reflection fallback

### PR DESCRIPTION
**Pull request by JamesNK**
Addresses the warning part of https://github.com/dotnet/aspnetcore/issues/45528

Adds a reflection fallback to invoking middleware. The goal is to avoid native AOT warnings on publish.

There is also a small performance benefit to using reflection instead of an expression with AOT. Avoids allocating and running expression interpreter. The ideal solution to performance here is using source generation to invoke middleware. That is still to be done before https://github.com/dotnet/aspnetcore/issues/45528 is complete.

### CODE REVIEW

**Review by davidfowl — APPROVED** [review]
LGTM

**Comment by JamesNK** [comment]
@eerhardt After making these changes locally, updating the assembly in the SDK folder, and republishing, I still see expressions related warnings.

```
/_/src/libraries/System.Linq.Expressions/src/System/Linq/Expressions/Interpreter/ArrayOperations.cs(23): AOT analysis warning IL3050: System.Linq.Expressions.Interpreter.NewArrayInitInstruction.Run(InterpretedFrame): Using member 'System.Array.CreateInstance(Type,Int32)' which has 'RequiresDynamicCodeAttribute' can break functionality when AOT compiling. The code for an array of the specified type might not be available. [C:\Users\james\source\repos\WebApplication25\WebApplication25\WebApplication25.csproj]
/_/src/libraries/System.Linq.Expressions/src/System/Linq/Expressions/Interpreter/ArrayOperations.cs(52): AOT analysis warning IL3050: System.Linq.Expressions.Interpreter.NewArrayInstruction.Run(InterpretedFrame): Using member 'System.Array.CreateInstance(Type,Int32)' which has 'RequiresDynamicCodeAttribute' can break functionality when AOT compiling. The code for an array of the specified type might not be available. [C:\Users\james\source\repos\WebApplication25\WebApplication25\WebApplication25.csproj]
/_/src/libraries/System.Linq.Expressions/src/System/Linq/Expressions/Interpreter/ArrayOperations.cs(87): AOT analysis warning IL3050: System.Linq.Expressions.Interpreter.NewArrayBoundsInstruction.Run(InterpretedFrame): Using member 'System.Array.CreateInstance(Type,Int32[])' which has 'RequiresDynamicCodeAttribute' can break functionality when AOT compiling. The code for an array of the specifiedtype might not be available. [C:\Users\james\source\repos\WebApplication25\WebApplication25\WebApplication25.csproj]
/_/src/libraries/System.Linq.Expressions/src/System/Dynamic/Utils/TypeUtils.cs(28): AOT analysis warning IL3050: System.Dynamic.Utils.TypeUtils.GetNullableType(Type): Using member 'System.Type.MakeGenericType(Type[])' which has 'RequiresDynamicCodeAttribute' can break functionality when AOT compiling. The native code for this instantiation might not be available at runtime. [C:\Users\james\source\repos\WebApplication25\WebApplication25\WebApplication25.csproj]
/_/src/libraries/System.Linq.Expressions/src/System/Linq/Expressions/MethodCallExpression.cs(1379): AOT analysis warning IL3050: System.Linq.Expressions.Expression.ApplyTypeArgs(MethodInfo,Type[]): Using member 'System.Reflection.MethodInfo.MakeGenericMethod(Type[])' which has 'RequiresDynamicCodeAttribute' can break functionality when AOT compiling. The native code for this instantiation might not be available at runtime. [C:\Users\james\source\repos\WebApplication25\WebApplication25\WebApplication25.csproj]
```

I mentioned on the original issue that it's odd the warnings are coming from `System.Linq.Expressions` files. Is the problem that  `System.Linq.Expressions` is not correctly annotated?

**Comment by davidfowl** [comment]
I thought [this was the issue](https://github.com/dotnet/linker/issues/27150).

**Comment by JamesNK** [comment]
This doesn't fix the expressions AOT warnings, but IMO this change is still worthwhile. Invoking via raw reflection is better than interpreting compiled expressions. And this change is pretty modest, self-contained, and is on the critical path (called potentially multiple times per request). Source gen is the best solution, but this is a good improvement for existing libraries who call `UseMiddleware<T>()`.

**Comment by davidfowl** [comment]
Yes this is fine. The source gen will also be usable in libraries.

File: src/Http/Http.Abstractions/src/Extensions/UseMiddlewareExtensions.cs [hunk 1]
```diff
@@ -107,7 +109,9 @@ public static class UseMiddlewareExtensions
                 return (RequestDelegate)invokeMethod.CreateDelegate(typeof(RequestDelegate), instance);
             }
 
-            var factory = Compile<object>(invokeMethod, parameters);
+            var factory = RuntimeFeature.IsDynamicCodeSupported
```
Comment (davidfowl) [inline src/Http/Http.Abstractions/src/Extensions/UseMiddlewareExtensions.cs:112]: @JamesNK can you make a change to this to use https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.runtimefeature.isdynamiccodecompiled?view=net-7.0 instead?

**Pull request merged**